src/contract_code.rs

```diff
@@ -0,0 +1,200 @@
+        Binary::from(bin).to_term(env)
+    }
+
+    fn open_bin<'a>(term: Term<'a>) -> NifResult<crate::Bytes> {
+        if !term.is_binary() {
+            Err(Error::BadArg)?;
+        }
+
+        let bin = Binary::from_term(term)?;
+        let data = bin.as_slice();
+        Ok(data.to_vec())
+    }
+
+    impl Encoder for crate::contract_code::Code {
+        fn encode<'a>(&self, env: Env<'a>) -> Term<'a> {
+            Term::map_from_pairs(
+                env,
+                &[
+                    (fields::type_info(), Term::list_new_empty(env)),
+                    (fields::byte_code(), make_bin(env, &self.byte_code)),
+                    (fields::payable(), self.payable.encode(env)),
+                    (fields::source_hash(), make_bin(env, &self.source_hash)),
+                    (
+                        fields::compiler_version(),
+                        make_bin(env, &self.compiler_version),
+                    ),
+                ],
+            )
+            .expect("Failed creating an Erlang map")
+        }
+    }
+
+    impl<'a> Decoder<'a> for crate::contract_code::Code {
+        fn decode(term: Term<'a>) -> NifResult<Self> {
+            if !term.map_get(fields::type_info())?.is_empty_list() {
+                Err(Error::BadArg)?;
+            }
+
+            let code = crate::contract_code::Code {
+                byte_code: open_bin(term.map_get(fields::byte_code())?)?,
+                payable: term.map_get(fields::payable())?.decode()?,
+                source_hash: open_bin(term.map_get(fields::source_hash())?)?,
+                compiler_version: open_bin(term.map_get(fields::compiler_version())?)?,
+            };
+            Ok(code)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn source_hash() {
+        let source = "contract Foo = ...";
+        // Taken from the original Erlang implementation
+        let expect = vec![
+            48, 58, 125, 237, 188, 44, 120, 213, 52, 155, 92, 4, 213, 8, 157, 236, 198, 161, 240,
+            9, 117, 91, 60, 167, 64, 44, 67, 82, 145, 174, 238, 243,
+        ];
+
+        assert_eq!(hash_source_code(&source), expect);
+    }
+
+    #[test]
+    fn sophia_contract_version3_serialize() {
+        let input = Code {
+            byte_code: "DUMMY_CODE".as_bytes().to_vec(),
+            source_hash: hash_source_code("contract Foo = ..."),
+            compiler_version: "3.1.4".as_bytes().to_vec(),
+            payable: true,
+        };
+        // Taken from the original Erlang implementation
+        let expect = vec![
+            246,70,3,160,48,58,125,237,188,44,120,213,52,155,92,4,213,8,157,236,198,161,
+            240,9,117,91,60,167,64,44,67,82,145,174,238,243,192,138,68,85,77,77,89,95,67,
+            79,68,69,133,51,46,49,46,52,1
+        ];
+
+        // TODO: figure out which is correct
+        // let expect = vec![
+        //     248,60,70,3,160,48,58,125,237,188,44,120,213,52,155,92,4,213,8,157,
```

Comment:
I figured it out. `type_info` is non-existent in FATE as it was used only in AEVM. The format however requires this field, for some reason. When I set the list to be empty, the serializations are in line with each other.

Reply:
@radrow I don't see that you have made any change here. This is still giving an output different from the one of the erlang implementation. Maybe you add a hardcoded serialization of type_info?

Reply:
Can you post what exactly you ran in erlang?

Reply:
Ah, sorry. Missed an important sentence. Type info is only for aevm, fate does not care about it. Although I see that we might need to accept a nonempty list here for compatibility reasons... I'd need to discuss it

Reply:
Since we are not going to use it for node sync yet, I am leaving it as it is. This field is an unnecessary residue and the compiler, which is what we are pursuing, should not use it anyway. Leaving a comment though so we don't forget.

Note that we are not rewriting all aeser features.